src/process/ProcessManagerImpl.cpp

```diff
@@ -133,26 +133,39 @@ ProcessManagerImpl::getNumRunningProcesses()
 
 ProcessManagerImpl::~ProcessManagerImpl()
 {
-    const auto killProcess = [&](ProcessExitEvent& pe) {
-        pe.mImpl->cancel(ABORT_ERROR_CODE);
-        forceShutdown(pe);
-    };
-    // Use SIGKILL on any processes we already used SIGINT on
-    while (!mKillable.empty())
-    {
-        auto impl = std::move(mKillable.front());
-        mKillable.pop_front();
-        killProcess(*impl);
-    }
-    // Use SIGKILL on any processes we haven't politely asked to exit yet
-    for (auto& pair : mProcesses)
+// First unregister our SIGCHLD handler; it has a raw pointer to us and
+// anyways we won't be able to bounce off ASIO for the work we're about
+// to do below.
 #ifndef _WIN32
     auto ec = ABORT_ERROR_CODE;
     mSigChild.cancel(ec);
 #endif
+
+    // Then trigger shutdown, if we haven't yet (it's idempotent). This will ask
+    // every running process to shut down politely and cancel all events,
+    // pending and running.
+    shutdown();
+
+    // At this point we're purely racing against process exits; we don't want to
+    // block or spin in a dtor by trying to trap SIGCHLD (and we might have
+    // dropped a SIGCHLD after cancelling the handler above anyways) but we can
+    // at least try to lose the race by sleeping briefly before calling waitpid
+    // below with WNOHANG in the handleSignalWait() call.
+    std::this_thread::sleep_for(std::chrono::milliseconds(10));
+
+    // Reap anything that politely shut down in response.
+    reapChildren();
+
+    // Then trigger forceful shutdown of everything that survived.
+    for (auto k : mKillable)
     {
-        killProcess(*pair.second);
+        forceShutdown(*k);
     }
+
+    // And reap all the processes that died after that, too, again trying to
+    // lose the race against such exits if we can.
+    std::this_thread::sleep_for(std::chrono::milliseconds(10));
+    reapChildren();
```

Comment:
I'm not 100% sure this sequence works as intended on Windows as process handling is done on the completion thread, but I think it might be "good enough" (force killing everything).
I did a few smoke tests, and things seems OK

Reply:
Yeah, I'm not sure it works _great_ anywhere. Ultimately we're in a bit of a bind as far as not wanting to block or fail, just do best-effort kills.

I think the great majority of our users (especially those running core subprocesses and killing them at random, eg. captive subprocesses of horizon) are going to be running in unix containers. And you have to be running a process when someone kills core to even hit this path. I expect this path is a no-op for most users most of the time.